program/rust/src/tests/test_ema.rs

```diff
@@ -1,23 +1,198 @@
 extern crate test_generator;
 
 use {
+    super::test_utils::update_clock_slot,
     crate::{
-        accounts::PriceAccount,
+        accounts::{
+            PriceAccount,
+            PythAccount,
+        },
+        c_oracle_header::{
+            PC_STATUS_TRADING,
+            PC_STATUS_UNKNOWN,
+            PC_VERSION,
+        },
+        deserialize::{
+            load_checked,
+            load_mut,
+        },
+        instruction::{
+            OracleCommand,
+            UpdPriceArgs,
+        },
         processor::{
             c_upd_aggregate,
             c_upd_twap,
+            process_instruction,
         },
+        tests::test_utils::AccountSetup,
     },
     bytemuck::Zeroable,
     csv::ReaderBuilder,
     serde::{
         Deserialize,
         Serialize,
     },
-    std::fs::File,
+    solana_program::pubkey::Pubkey,
+    std::{
+        fs::File,
+        mem::size_of,
+    },
     test_generator::test_resources,
 };
 
+#[test]
+fn test_ema_multiple_publishers_same_slot() -> Result<(), Box<dyn std::error::Error>> {
+    let mut instruction_data = [0u8; size_of::<UpdPriceArgs>()];
+    populate_instruction(&mut instruction_data, 10, 1, 1);
+
+    let program_id = Pubkey::new_unique();
+
+    let mut funding_setup = AccountSetup::new_funding();
+    let funding_account = funding_setup.as_account_info();
+
+    let mut price_setup = AccountSetup::new::<PriceAccount>(&program_id);
+    let mut price_account = price_setup.as_account_info();
+    price_account.is_signer = false;
+    PriceAccount::initialize(&price_account, PC_VERSION).unwrap();
+
+    {
+        let mut price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        price_data.num_ = 1;
+        price_data.comp_[0].pub_ = *funding_account.key;
+    }
+
+    let mut clock_setup = AccountSetup::new_clock();
+    let mut clock_account = clock_setup.as_account_info();
+    clock_account.is_signer = false;
+    clock_account.is_writable = false;
+
+    update_clock_slot(&mut clock_account, 1);
+
+    process_instruction(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone(),
+        ],
+        &instruction_data,
+    )?;
+
+    {
```

Comment:
agreed with what you said, added a new helper method `add_publisher` but for `update_publisher_price` I don't think abstracting it is going to make it any easier to understand since it's already pretty straightforward, I think the reason its hard to understand here is because the first `populate_instruction()` is pretty high above, will move it below right before `process_instruction`

Reply:
also for `update_publisher_price` this is existing code, and we can refactor them in a following PR, what do you think?